command_test.go

```diff
@@ -1079,6 +1080,192 @@ func TestHelpExecutedOnNonRunnableChild(t *testing.T) {
+	}
+
+	if !helpCalled {
+		t.Error("Overridden help function not called")
+	}
+}
+
+func TestHelpOverrideOnRootWithChildAndFlags(t *testing.T) {
+	var helpCalled bool
+	rootCmd := &Command{Use: "root"}
+	subCmd := &Command{Use: "child"}
+	rootCmd.AddCommand(subCmd)
+
+	var myFlag bool
+	subCmd.Flags().BoolVar(&myFlag, "myflag", false, "")
+
+	rootCmd.SetHelpFunc(func(c *Command, args []string) {
+		helpCalled = true
+		if c.Name() != "child" {
+			t.Errorf(`Expected command name: "child", got %q`, c.Name())
+		}
+		if len(args) != 2 || args[0] != "arg1" || args[1] != "arg2" {
+			t.Errorf("Expected args [args1 arg2], got %v", args)
+		}
+	})
+
+	_, err := executeCommand(rootCmd, "child", "arg1", "--myflag", "arg2", "--help")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	if !helpCalled {
+		t.Error("Overridden help function not called")
+	}
+}
+
+func TestHelpOverrideOnRootWithChildAndFlagsButParsingDisabled(t *testing.T) {
+	var helpCalled bool
+	rootCmd := &Command{Use: "root"}
+	subCmd := &Command{Use: "child", DisableFlagParsing: true}
+	rootCmd.AddCommand(subCmd)
+
+	var myFlag bool
+	subCmd.Flags().BoolVar(&myFlag, "myflag", false, "")
+
+	rootCmd.SetHelpFunc(func(c *Command, args []string) {
+		helpCalled = true
+		if c.Name() != "child" {
+			t.Errorf(`Expected command name: "child", got %q`, c.Name())
+		}
+		if len(args) != 4 ||
+			args[0] != "arg1" || args[1] != "--myflag" || args[2] != "arg2" || args[3] != "--help" {
+			t.Errorf("Expected args [args1 --myflag arg2 --help], got %v", args)
+		}
+	})
+
+	_, err := executeCommand(rootCmd, "child", "arg1", "--myflag", "arg2", "--help")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	if !helpCalled {
+		t.Error("Overridden help function not called")
+	}
+}
+
+func TestHelpCommandOverrideOnChild(t *testing.T) {
+	var helpCalled bool
+	rootCmd := &Command{Use: "root"}
+	subCmd := &Command{Use: "child"}
+	rootCmd.AddCommand(subCmd)
+
+	subCmd.SetHelpFunc(func(c *Command, args []string) {
+		helpCalled = true
+		if c.Name() != "child" {
+			t.Errorf(`Expected command name: "child", got %q`, c.Name())
+		}
+		if len(args) != 2 || args[0] != "arg1" || args[1] != "arg2" {
+			t.Errorf("Expected args [args1 arg2], got %v", args)
+		}
+	})
+
+	_, err := executeCommand(rootCmd, "help", "child", "arg1", "arg2")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	if !helpCalled {
+		t.Error("Overridden help function not called")
+	}
+}
+
+func TestHelpCommandOverrideOnRootWithChild(t *testing.T) {
+	var helpCalled bool
+	rootCmd := &Command{Use: "root"}
+	subCmd := &Command{Use: "child"}
+	rootCmd.AddCommand(subCmd)
+
+	rootCmd.SetHelpFunc(func(c *Command, args []string) {
+		helpCalled = true
+		if c.Name() != "child" {
+			t.Errorf(`Expected command name: "child", got %q`, c.Name())
+		}
+		if len(args) != 2 || args[0] != "arg1" || args[1] != "arg2" {
+			t.Errorf("Expected args [args1 arg2], got %v", args)
+		}
+	})
```

Comment:
We repeat the code of the help function many times, maybe add a helper struct keeping the , so we can do:

     h := helper{t: t, command: "child", args: []string{"args1", "arg2"}}
     rootCmd.SetHelpFunc(h.helpFunc)
     _, err := executeCommand(rootCmd, "help", "child", "arg1", "arg2")

     if !h.helpCalled {
         ...
     }

Reply:
I like it!
I tried to do what you suggest.